File: tests/python/test_vosc.py
"""
Python binding tests for VOSC indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

# Add parent directory to path to import the built module
sys.path.insert(0, str(Path(__file__).parent.parent.parent / 'target/wheels'))

try:
    import my_project as ta_indicators
except ImportError:
    pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS


class TestVosc:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()
    
    def test_vosc_accuracy(self, test_data):
        """Test VOSC matches expected values from Rust tests"""
        # Extract volume data
        volume = test_data.volume
        
        # Parameters from Rust test
        short_period = 2
        long_period = 5
        
        # Run VOSC calculation
        result = ta_indicators.vosc(volume, short_period, long_period)
        
        # Expected values from Rust test
        expected_last_five = [
            -39.478510754298895,
            -25.886077312645188,
            -21.155087549723756,
            -12.36093768813373,
            48.70809369473075,
        ]
        
        # Verify output
        assert len(result) == len(volume), "VOSC length mismatch"
        
        # Check last 5 values
        for i, expected in enumerate(expected_last_five):
            assert_close(result[-(5-i)], expected, 1e-1)
        
        # Check warmup period (should be NaN)
        warmup_period = long_period - 1
        for i in range(warmup_period):
            assert np.isnan(result[i]), f"Expected NaN at index {i}"
    
    def test_vosc_errors(self):
        """Test error handling"""
        # Test empty data
        with pytest.raises(Exception):
            ta_indicators.vosc([], 2, 5)
        
        # Test zero period
        with pytest.raises(Exception):
            ta_indicators.vosc([1.0, 2.0, 3.0], 0, 5)
        
        # Test short period > long period
        with pytest.raises(Exception):
            ta_indicators.vosc([1.0, 2.0, 3.0, 4.0, 5.0], 5, 2)
        
        # Test all NaN values
        with pytest.raises(Exception):
            ta_indicators.vosc([np.nan, np.nan, np.nan], 2, 3)
    
    def test_vosc_batch(self, test_data):
        """Test VOSC batch calculation"""
        volume = test_data.volume
        
        # Run batch calculation with ranges
        result = ta_indicators.vosc_batch(
            volume,
            (2, 10, 1),  # short_period_range
            (5, 20, 1)   # long_period_range
        )
        
        # Check result structure
        assert 'values' in result
        assert 'short_periods' in result
        assert 'long_periods' in result
        
        # Verify dimensions
        n_combos = len(result['short_periods'])
        assert result['values'].shape == (n_combos, len(volume))
    
    def test_vosc_stream(self):
        """Test VOSC streaming calculation"""
        # Create stream
        stream = ta_indicators.VoscStream(2, 5)
        
        # Test data
        test_values = [100.0, 120.0, 110.0, 130.0, 125.0, 140.0, 135.0]
        results = []
        
        # Update stream
        for value in test_values:
            result = stream.update(value)
            results.append(result)
        
        # First few values should be None (warmup)
        assert all(r is None for r in results[:4])
        
        # After warmup, should get values
        assert results[4] is not None


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: tests/wasm/test_vosc.js
/**
 * WASM binding tests for VOSC indicator.
 * These tests mirror the Rust unit tests to ensure WASM bindings work correctly.
 */
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { createRequire } from 'module';
import { 
    loadTestData, 
    assertArrayClose, 
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS 
} from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {
    // Load a known-good NodeJS wrapper to avoid ESM/"env" import issues
    try {
        const req = createRequire(import.meta.url);
        // Use the full CommonJS wrapper copy to avoid ESM 'module' issues
        const mod = req(path.join(__dirname, 'my_project_full.cjs'));
        wasm = mod; // CommonJS exports
    } catch (error) {
        console.error('Failed to load local WASM wrapper (tests/wasm/my_project.js).');
        console.error('If missing, run: wasm-pack build -- --features wasm --no-default-features');
        throw error;
    }

    testData = loadTestData();
});

test('VOSC accuracy', () => {
    // Test with default parameters from Python test
    const volume = testData.volume;
    const shortPeriod = 2;
    const longPeriod = 5;
    
    // Run VOSC calculation
    const result = wasm.vosc_js(volume, shortPeriod, longPeriod);
    
    // Expected values from Python test
    const expectedLastFive = [
        -39.478510754298895,
        -25.886077312645188,
        -21.155087549723756,
        -12.36093768813373,
        48.70809369473075,
    ];
    
    // Verify output length
    assert.strictEqual(result.length, volume.length, 'VOSC length mismatch');
    
    // Check last 5 values
    const startIdx = result.length - 5;
    for (let i = 0; i < expectedLastFive.length; i++) {
        assertClose(result[startIdx + i], expectedLastFive[i], 0.1, 
            `VOSC mismatch at index ${startIdx + i}`);
    }
    
    // Check warmup period (should be NaN)
    const warmupPeriod = longPeriod - 1;
    for (let i = 0; i < warmupPeriod; i++) {
        assert.ok(isNaN(result[i]), `Expected NaN at index ${i}`);
    }
});

test('VOSC error handling', () => {
    // Test empty data
    assert.throws(() => {
        wasm.vosc_js([], 2, 5);
    }, 'Should throw on empty data');
    
    // Test zero short period
    assert.throws(() => {
        wasm.vosc_js([1.0, 2.0, 3.0], 0, 5);
    }, 'Should throw on zero short period');
    
    // Test zero long period
    assert.throws(() => {
        wasm.vosc_js([1.0, 2.0, 3.0], 2, 0);
    }, 'Should throw on zero long period');
    
    // Test short period > long period
    assert.throws(() => {
        wasm.vosc_js([1.0, 2.0, 3.0, 4.0, 5.0], 5, 2);
    }, 'Should throw when short period > long period');
    
    // Test all NaN values
    assert.throws(() => {
        wasm.vosc_js([NaN, NaN, NaN], 2, 3);
    }, 'Should throw on all NaN values');
});

test('VOSC fast API', () => {
    const volume = testData.volume;
    const shortPeriod = 2;
    const longPeriod = 5;
    
    // Allocate memory
    const len = volume.length;
    const inPtr = wasm.vosc_alloc(len);
    const outPtr = wasm.vosc_alloc(len);
    
    try {
        // Copy data to WASM memory
        const wasmMemory = new Float64Array(wasm.__wasm.memory.buffer);
        wasmMemory.set(volume, inPtr / 8);
        
        // Run calculation
        wasm.vosc_into(inPtr, outPtr, len, shortPeriod, longPeriod);
        
        // Read results
        const result = new Float64Array(wasm.__wasm.memory.buffer, outPtr, len);
        const resultCopy = Array.from(result);
        
        // Compare with safe API
        const safeResult = wasm.vosc_js(volume, shortPeriod, longPeriod);
        assertArrayClose(resultCopy, safeResult, 1e-10, 
            'Fast API should match safe API');
    } finally {
        // Clean up
        wasm.vosc_free(inPtr, len);
        wasm.vosc_free(outPtr, len);
    }
});

test('VOSC in-place operation', () => {
    const volume = testData.volume;
    const shortPeriod = 2;
    const longPeriod = 5;
    
    // Allocate memory and copy data
    const len = volume.length;
    const ptr = wasm.vosc_alloc(len);
    
    try {
        const wasmMemory = new Float64Array(wasm.__wasm.memory.buffer);
        wasmMemory.set(volume, ptr / 8);
        
        // Run in-place calculation
        wasm.vosc_into(ptr, ptr, len, shortPeriod, longPeriod);
        
        // Read results
        const result = new Float64Array(wasm.__wasm.memory.buffer, ptr, len);
        const resultCopy = Array.from(result);
        
        // Compare with safe API
        const safeResult = wasm.vosc_js(volume, shortPeriod, longPeriod);
        assertArrayClose(resultCopy, safeResult, 1e-10, 
            'In-place operation should match safe API');
    } finally {
        wasm.vosc_free(ptr, len);
    }
});

test('VOSC batch API', () => {
    const volume = testData.volume;
    
    // Test batch calculation with small ranges
    const config = {
        short_period_range: [2, 4, 1],
        long_period_range: [5, 7, 1]
    };
    
    const result = wasm.vosc_batch(volume, config);
    
    // Verify structure
    assert.ok(result.values, 'Batch result should have values');
    assert.ok(result.combos, 'Batch result should have combos');
    assert.ok(result.rows > 0, 'Batch result should have rows');
    assert.strictEqual(result.cols, volume.length, 'Batch cols should match data length');
    
    // Verify dimensions
    const expectedCombos = 3 * 3; // 3 short periods * 3 long periods (filtered for valid combos)
    assert.ok(result.combos.length > 0, 'Should have valid combinations');
    assert.strictEqual(result.values.length, result.rows * result.cols, 
        'Values array size should match dimensions');
});

test.after(() => {
    console.log('VOSC WASM tests completed');
});
